### PR TITLE
[ARM/Linux] Fix arm-softfp calling convention

### LINKED ISSUE #4745: ARM/Linux "structinregs" unit test fail
Unit test `JIT.Methodical.structs.systemvbringup.structinregs` fails in ARM/Linux.

```
...
Native S1: 1, 2, 3, 4
S1: 1, 2, 3, 4
S1R: 1, 2, 3, 4
Native S2: 1, 2, 3.000000
S2: 1, 2, 3
S2R: 1, 2, 3
Native S3: 1, 2, 3.000000
S3: 1, 2, 3
S3R: 1, 2, 3
Native S4: 1073741824, 157501122515305666310423183360.000000

Program received signal SIGSEGV, Segmentation fault.
0x00000000 in ?? ()
(gdb) bt
#0  0x00000000 in ?? ()
#1  0x6ffe75e0 in InvokeCallback4R (callback=0x1, s=...) at structinregs.cpp:367
#2  0x6ffd6860 in ?? ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

P/Invoke to `InvokeCallback4R()` in `structpinvoketests.cs` fails. While `InvokeCallback{1-3}R()` works fine.

### PR DESCRIPTION
**Pull request by wateret**
Fix arm-softfp calling convention not to follow HFA which is only for hardfp.

Fix #4745 for arm-softfp

### CODE REVIEW

**Comment by wateret** [comment]
Details are https://github.com/dotnet/coreclr/issues/4745#issuecomment-222598653.

**Comment by wateret** [comment]
cc: @papaslavik, @janvorli, @parjong

**Comment by wateret** [comment]
@papaslavik I agree with #4930 and  I think it needs one more change in `impFixupCallStructReturn` - caller side.

**Comment by janvorli** [comment]
cc: @dotnet/jit-contrib

File: src/jit/importer.cpp [hunk 1]
```diff
@@ -6958,7 +6958,7 @@ bool  Compiler::impMethodInfo_hasRetBuffArg(CORINFO_METHOD_INFO * methInfo)
     //
   #ifdef FEATURE_HFA
     // On ARM HFAs are returned in registers.
-    if (!info.compIsVarArgs && IsHfa(methInfo->args.retTypeClass))
+    if (!info.compIsVarArgs && !opts.compUseSoftFP && IsHfa(methInfo->args.retTypeClass))
     {
```
Comment (CarolEidt) [inline src/jit/importer.cpp]: Would it make more sense to modify the definition of IsHfa to do the check for opts.compUseSoftFP?

Comment (briansull) [inline src/jit/importer.cpp]: I think that you should try changing IsHfa to turn off all Hfa's foft compUseSoftFP

bool Compiler::IsHfa(CORINFO_CLASS_HANDLE hClass)
{
#ifdef FEATURE_HFA
    return varTypeIsFloating(GetHfaType(hClass));
#else
    return false;
#endif
}

Comment (briansull) [inline src/jit/importer.cpp]: For now you can modify the above method to return false when compUseSoftFP is enabled,
But ultimately the ifdef FEATURE_HFA might need to be undef-ed when building the VM and JIT.
Currently the VM marks certain structs as HFA's and may treat them differently in Interop code.

Comment (papaslavik) [inline src/jit/importer.cpp]: @briansull let me ask why that's not done for var args then?
Is the IsHfa just about arguments?

**Comment by wateret** [comment]
@CarolEidt @briansull
I first tried to check `compUseSoftFP` inside `IsHfa` but there were problems with e.g. there was assertion failure in code like `assert(IsHfa(...))`. So I thought it may make more sense to undef `FEATURE_HFA` and seems like it works. 

I did not run the entire unit test yet but ran some unit tests that are closely related to these changes. 
All tests in `JIT/jit64/hfa` pass(8 fails before) and one more test in `JIT/SIMD` passes than before.

File: src/jit/importer.cpp (hunk 1 above)
Comment (briansull) [inline src/jit/importer.cpp]: This method Compiler::IsHfa answers the question is this class handle an HFA type.

For a varargs method the calling convention says to ignore HFA and pass them as if they weren't HFA's
And you can have a mix of methods some which are varargs and some that are not varargs.

But I think that with opts.compUseSoftFP you can't have this kind of mix of methods.
Basically if opts.compUseSoftFP is set when compiling a method it has to be on when compiling every method, 
as you are modifying the calling convention and both the called method and the callsite both have to agree how to pass and receive arguments.

My belief is that once compUseSoftFP is set it should always be set and you no longer have any HFA types at all.

Comment (papaslavik) [inline src/jit/importer.cpp]: So, you mean that in the same function we can have calls to varargs and non-varargs. Yes, I didn't think about that.

But do you mean that Compiler::IsHfa matters only for calling convention and we won't get anything else wrong if we switch it off for ARM softfp calling convention?

Comment (briansull) [inline src/jit/importer.cpp]: Yes, When a struct is an HFA it effects how it is passed and returned.
It doesn't really effect anything else about the struct.

**Comment by wateret** [comment]
@briansull Please take a look again.

**Comment by briansull** [comment]
LGTM

**Pull request merged**